sshutil/options.go

```diff
@@ -0,0 +1,90 @@
+package sshutil
+
+import (
+	"bytes"
+	"encoding/base64"
+	"io/ioutil"
+	"text/template"
+
+	"github.com/Masterminds/sprig/v3"
+	"github.com/pkg/errors"
+	"go.step.sm/crypto/internal/step"
+)
+
+func getFuncMap(failMessage *string) template.FuncMap {
+	m := sprig.TxtFuncMap()
+	m["fail"] = func(msg string) (string, error) {
+		*failMessage = msg
+		return "", errors.New(msg)
+	}
+	return m
+}
+
+// Options are the options that can be passed to NewCertificate.
+type Options struct {
+	CertBuffer *bytes.Buffer
+}
+
+func (o *Options) apply(cr CertificateRequest, opts []Option) (*Options, error) {
+	for _, fn := range opts {
+		if err := fn(cr, o); err != nil {
+			return o, err
+		}
+	}
+	return o, nil
+}
+
+// Option is the type used as a variadic argument in NewCertificate.
+type Option func(cr CertificateRequest, o *Options) error
+
+// WithTemplate is an options that executes the given template text with the
+// given data.
+func WithTemplate(text string, data TemplateData) Option {
+	return func(cr CertificateRequest, o *Options) error {
+		terr := new(TemplateError)
+		funcMap := getFuncMap(&terr.Message)
+
+		tmpl, err := template.New("template").Funcs(funcMap).Parse(text)
+		if err != nil {
+			return errors.Wrapf(err, "error parsing template")
+		}
+
+		buf := new(bytes.Buffer)
+		data.SetCertificateRequest(cr)
+		if err := tmpl.Execute(buf, data); err != nil {
+			if terr.Message != "" {
```

Comment:
Can you explain why we do this trickery with the error (rather than just return the error from Execute). I presume the error returned by execute has very little information.

Reply:
And by explain, I just mean 'leave a comment'.

Reply:
Sure, the error is very obscured, by text/template, so I just want to return the string in the template.